x/wasm/keeper/msg_dispatcher.go

```diff
@@ -147,6 +148,22 @@ func (d MessageDispatcher) DispatchSubmessages(ctx sdk.Context, contractAddr sdk
 		} else {
 			// Issue #759 - we don't return error string for worries of non-determinism
 			moduleLogger(ctx).Info("Redacting submessage error", "cause", err)
+
+			// Emiting the errors to provide useful information to the end-users
+			// see: https://github.com/CosmWasm/wasmd/issues/1122
+			msgType := getMsgType(msg.Msg)
+			attrs := []sdk.Attribute{
+				sdk.NewAttribute(types.AttributeKeyContractAddr, contractAddr.String()),
+				sdk.NewAttribute(types.AttributeKeySubMsgError, err.Error()),
+				sdk.NewAttribute(types.AttributeKeyMsgtype, msgType),
+			}
+			if msgType == "Stargate" {
+				attrs = append(attrs, sdk.NewAttribute(types.AttributeKeyStargateTypeUrl, msg.Msg.Stargate.TypeURL))
+			}
+
+			event := sdk.NewEvent(types.CustomContractEventPrefix+types.EventTypeRedactedError, attrs...)
+			ctx.EventManager().EmitEvent(event)
```

Comment:
Please note that this is a cached context. In the SDK events are only emitted when the TX succeeds. There is no guarantee that this is the case. 
The contract can decide to fail the TX on reply and rollback. When the contract handles the error reply, is there still value in the event?

Reply:
Ahh good catch! I'll make sure those events are copied when if the context is discarded. 

It's also a good question what we want to happen if the failed message is caught by the contract. I started writing this to be able to provide feedback when a transaction does ultimately fails, so I think it's ok for the events to be omitted if the contract catches the failure, as I assume that means the contract is properly handling the errors. 

Do you think there is value in surfacing those errors anyway?

Reply:
ahh, this is annoying: this would have to be done in baseapp:runTx. Trying to figure out a good workaround